velox/functions/lib/Re2Functions.cpp

```diff
@@ -496,46 +496,119 @@ bool matchExactPattern(
 }
 
-std::pair<bool, int32_t> matchRelaxedFixedForwardAscii(
+namespace {
+class RelaxedPatternMatchResult {
+ public:
+  static RelaxedPatternMatchResult success(size_t cursor) {
+    return RelaxedPatternMatchResult{true, cursor, -1};
+  }
+
+  static RelaxedPatternMatchResult failed(
+      size_t cursor,
+      int32_t failingPatternIndex) {
+    return RelaxedPatternMatchResult{false, cursor, failingPatternIndex};
+  }
+
+  static RelaxedPatternMatchResult failedForTooShortInput() {
+    return RelaxedPatternMatchResult{false, 0, -1};
+  }
+
+  bool match() {
+    return match_;
+  }
+
+  // The input the not long enough to match.
+  bool tooShortInputToMatch() {
+    return failingPatternIndex_ < 0;
+  }
+
+  size_t cursor() {
+    return cursor_;
+  }
+
+  int32_t failingPatternIndex() {
+    return failingPatternIndex_;
+  }
+
+ private:
+  RelaxedPatternMatchResult(
+      bool match,
+      size_t cursor,
+      int32_t failingPatternIndex)
+      : match_(match),
+        cursor_(cursor),
+        failingPatternIndex_(failingPatternIndex) {}
+
+  // Whether it matches.
+  bool match_;
+
+  // If match == true, it stores the cursor in the input
+  // when we finished matching.
+  // If match == false, it stores the cursor in the input where we failed the
+  // matching.
+  size_t cursor_;
+
+  // If match == false, this field is used. It is assigned -1 if the match
+  // fails not because of any sub-pattern(e.g. the input is too short).
+  int32_t failingPatternIndex_;
+};
+} // namespace
+
+// Matches a kRelaxedFixed pattern forward against ASCII input.
+RelaxedPatternMatchResult matchRelaxedFixedForwardAscii(
     StringView input,
     const PatternMetadata& patternMetadata,
-    size_t start) {
-  if (input.size() - start < patternMetadata.length()) {
-    return std::make_pair(false, -1);
+    size_t start,
+    size_t startSubPatternIndex,
+    size_t endSubPatternIndex) {
+  const auto lengthOfThePattern =
+      patternMetadata.lengthOf(startSubPatternIndex, endSubPatternIndex);
+  // Compare the length first.
+  if (input.size() - start < lengthOfThePattern) {
+    return RelaxedPatternMatchResult::failedForTooShortInput();
   }
 
-  for (const auto& subPattern : patternMetadata.subPatterns()) {
+  const auto patternOffset =
+      patternMetadata.subPatterns()[startSubPatternIndex].start;
+  for (auto i = startSubPatternIndex; i < endSubPatternIndex + 1; ++i) {
+    const auto& subPattern = patternMetadata.subPatterns()[i];
+    const auto currentMatchingOffset =
+        start + (subPattern.start - patternOffset);
     if (subPattern.kind == SubPatternKind::kLiteralString &&
         std::memcmp(
-            input.data() + start + subPattern.start,
+            input.data() + currentMatchingOffset,
             patternMetadata.fixedPattern().data() + subPattern.start,
             subPattern.length) != 0) {
-      return std::make_pair(false, -1);
+      return RelaxedPatternMatchResult::failed(currentMatchingOffset, i);
```

Comment:
Perhaps, we could have a shortcut here that returns failed() if the remaining number of bytes is exactly lengthOfThePattern (input.size() - start == lengthOfThePattern)